### PR TITLE
Fixing `set_start_value` only accept numbers (#2237)

### PR DESCRIPTION
**Pull request by henriquebecker91**
Intends to close #2237 and some related open issues (#649, #794, #853).

Some guidance is needed. For now, I just edited what I found to be the relevant piece of documentation, and added a new tests as well flags to an old test (allowing more granular selection of which attributes are tested). I would like to know:

1. There are other places where the documentation needs to be changed? `JuMP` probably will need a change, but there are other points inside MOI?
2. Seems to me that allowing `nothing` to be passed is something done at the wrapper level, and the generic methods in MOI already allow it. The test will probably force the wrappers to allow this behavior.
3. How is the best way to check if the test is working? Should I download a bunch of wrappers and make they use my version of MOI, and then call their tests?

No hurry, @odow, I do not know if I will work during the week, but I will try finish it next weekend if I am sure of what needs to be done yet.

### CODE REVIEW

File: src/attributes.jl
Comment (mlubin) [inline src/attributes.jl:969]: The `VariablePrimalStart` is a hint that the optimizer "may" use to warm-start the solve when it's set. It's redundant to say the solver "is free to decide which value to use" when it's set to `nothing`, because this is also true when it's set to a number. Just say "May be a number or `nothing` (unset)."

**Comment by odow** [comment]
1. We should update the documentation for `ConstraintDualStart` and `ConstraintPrimaStart` as well:
https://github.com/jump-dev/MathOptInterface.jl/blob/a6274e63f8cc020e16cb99a0b2d52bd8062e1613/src/attributes.jl#L1046-L1053

2. Rather than modifying the current test, you should add new tests
```Julia
function test_variable_primal_start(model, config)
    MOI.empty!(model)
    x = MOI.add_variable(model)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) === nothing
    MOI.set(model, MOI.VariablePrimalStart(), x, 1.0)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) == 1.0
    MOI.set(model, MOI.VariablePrimalStart(), x, nothing)
    @test MOI.get(model, MOI.VariablePrimalStart(), x) === nothing
end
```

3. You can add a test for the test here: https://github.com/jump-dev/MathOptInterface.jl/blob/master/test/Test/modellike.jl
Something like
```julia
model = MOIU.MockOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()))
MOI.Test.test_variable_primal_start(model, MOI.Test.TestConfig())
```

**Comment by henriquebecker91** [comment]
@odow, you asked to do the same modification of documentation to `ConstraintPrimalStart` and `ConstraintDualStart`, ok, I did (not pushed yet), but then you suggest creating just a `function test_variable_primal_start(model, config)`, tests should be added for `ConstraintPrimalStart` and `ConstraintDualStart`, no?

**Comment by henriquebecker91** [comment]
I made a new commit. It updates the documentation of `VariablePrimalStart`, `ConstraintDualStart`, and `ConstraintPrimalStart` to what was suggested by @mlubin.

I also reverted the test I modified and instead created a new test, but not only for `VariablePrimalStart` (also for the constraint start values), and it is a little more involved than the test suggested by @odow.

I call the new test in the fashion suggested by @odow. The `config` parameter is not used, should it be used? 

My new test method has 5 commented `@test` calls. They fail if uncommented. I am not sure if this is desired behavior or not. I expect a decision from the maintainers about removing or uncommenting them. In the current situation (with these 5 commented) all tests pass.

**Comment by henriquebecker91** [comment]
Should I be doing something here?

**Comment by henriquebecker91** [comment]
Right now, I have a window of opportunity to make any changes necessary for this to be approved, while I wait for some reviews of my last paper. Any directions?

File: src/attributes.jl
Comment (blegat) [inline src/attributes.jl:967]: nothing -> `nothing`

File: src/Test/modellike.jl
Comment (blegat) [inline src/Test/modellike.jl:447]: I would expect it to be `nothing`, not `[nothing]`

File: src/Test/modellike.jl
Comment (blegat) [inline src/Test/modellike.jl:450]: Why is this commented ?

Comment (henriquebecker91) [reply src/Test/modellike.jl:450]: As I said above:

> My new test method has 5 commented @test calls. They fail if uncommented. I am not sure if this is desired behavior or not. I expect a decision from the maintainers about removing or uncommenting them. In the current situation (with these 5 commented) all tests pass.

File: src/Test/modellike.jl
Comment (henriquebecker91) [reply src/Test/modellike.jl:447]: I will check this. I noticed now that I do not query the `cpattr` of `c` before setting it the first time. If the initial value of `cpattr` for `c` is `[nothing]`, then it would make sense keep it that way, to avoid breaking the reasoning that initially motivated this behavior, that is, ``if variables with an attribute unset return `nothing` when queried, then setting them with the returned value (i.e., nothing) should restore the original/unset state''. 

I have to admit I did not deal directly with this kinda of constraint, so I do not know what to expect here.
```julia
c = MOI.add_constraint(model, MOIU.operate(vcat, Float64, 2.0fz + 1.0), MOI.Nonnegatives(1))
```

File: src/attributes.jl
Comment (henriquebecker91) [reply src/attributes.jl:967]: Already fixed in my local copy. Will be together with the fix of the other problems.

File: src/Test/modellike.jl
Comment (henriquebecker91) [reply src/Test/modellike.jl:447]: Oh, okay, I was already checking it before and it has value `nothing`. So the `MOI.set` should take `nothing` and `MOI.get` should return `nothing`. I changed the tests and it works. Should we have some tests for `nothing` values inside the arrays that may be used to set this attribute, or setting the whole constraint to `nothing` and getting `nothing` back is enough?

File: src/Test/modellike.jl
Comment (henriquebecker91) [reply src/Test/modellike.jl:450]: Expanding a little on the above:

The documentation for [`ListOfVariableAttributesSet`](https://jump.dev/MathOptInterface.jl/stable/apireference/#MathOptInterface.ListOfVariableAttributesSet) and for [`ListOfConstraintAttributesSet`](https://jump.dev/MathOptInterface.jl/stable/apireference/#MathOptInterface.ListOfConstraintAttributesSet) are both not very clear (at least to me) on what happens if the attributes are unset.

* If I search for the attributes in these lists from a newly created model, they do not exist in the lists, as it is expected.
* If I search for them in these lists after setting them  in the model, they do exist in the lists, as it is expected.
* The problem is that after *unsetting* the attributes, they keep existing in these lists, I do not know if this is desired behaviour.

More specifically, I do not know if this is a list for every attribute that was ever set (i.e., `MOI.set` was called with it), or just for the attributes with a value different from `nothing` at the moment the list is queried.

Comment (blegat) [reply src/Test/modellike.jl:450]: Yes, this is a bit ambiguous, we don't want the definition to be to constrained here as it would make it difficult to implement and it does not change anything for the main use case which is using it for copying.
You can remove these tests or add a comment explaining that it currently fails but we may want to enable these in the future.

File: src/Test/modellike.jl
Comment (henriquebecker91) [reply src/Test/modellike.jl]: Just changed the tests to expect `nothing` and set `nothing`. No test for `nothing` elements inside a Vector.

File: src/Test/modellike.jl
Comment (henriquebecker91) [reply src/Test/modellike.jl:461]: I kept them commented and added a comment explaining why they are commented referencing this thread.

**Comment by henriquebecker91** [comment]
I followed all @blegat suggestions and did a minor cosmetic change (changed `!(... \in ...)` to `... \notin ...`).

**Comment by henriquebecker91** [comment]
Screenshot of tests passing:
[](https://postimg.cc/cKW5pLBj)

File: src/Test/modellike.jl
Comment (blegat) [reply src/Test/modellike.jl]: It's ok if `nothing` element inside vector for a vector constraint is undefined behavior for now, this is out of the scope of this PR.

File: src/Test/modellike.jl
Comment (blegat) [inline src/Test/modellike.jl:454]: unsetted -> unset

**Review by blegat — APPROVED** [review]
Can you remove the "Draft" state of the PR ?

**Comment by henriquebecker91** [comment]
Sorry, forgot about that. Now it is ready for review. I will make that small comment change.

File: src/attributes.jl
Comment (henriquebecker91) [reply src/attributes.jl]: Done.

**Pull request merged**
